### PR TITLE
graphql basic request parsing and error management

### PR DESCRIPTION
**Pull request by tardyp**
Some basic infrastructure.

we had api/v2

graphql will be at api/v3

Goal is to let the user enable it via `c['www']['graphql'] = {'debug'=True}`

we make extra effort to not require graphql-core unless enabled

### CODE REVIEW

File: master/buildbot/test/unit/www/test_graphql.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        yield self.render_resource(self.rsrc, b"/", method=b"PATCH")
+        self.assertSimpleError("invalid HTTP method", 400)
+
+    @defer.inlineCallbacks
+    def test_get_query(self):
+        yield self.render_resource(
+            self.rsrc,
+            b"/?query={tests{id}}",
+        )
+        # TODO verify results
```
Comment (p12tic) [inline master/buildbot/test/unit/www/test_graphql.py:70]: WIP? Should be easy to capture what the current call returns and just assert that this does not change.

File: master/buildbot/www/graphql.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    def reconfigResource(self, new_config):
+        # @todo v2 has cross origin support, which might need to be factorized
+        self.gql_config = new_config.www.get("graphql")
+        self.debug = True
+        self.graphql = None
+        if self.gql_config is not None:
+            try:
+                import graphql
+            except ImportError:
```
Comment (p12tic) [inline master/buildbot/www/graphql.py:43]: Needs `pragma: no cover`.

File: master/buildbot/www/graphql.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            raise Error(400, [e.formatted for e in errors])
+
+    @defer.inlineCallbacks
+    def asyncRender(self, request):
+        if self.graphql is None:
+            raise Error(501, "graphql not enabled")
+
+        # graphql accepts its query either in post data or get's query
+        if request.method == b"POST":
+            query = request.content.read()
```
Comment (p12tic) [inline master/buildbot/www/graphql.py:82]: Could we add a test for this?

File: master/buildbot/www/graphql.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        elif request.method in (b"GET"):
+            if b"query" not in request.args:
+                raise Error(400, b"GET request must contain a 'query' parameter")
```
Comment (p12tic) [inline master/buildbot/www/graphql.py:85]: Could we add a test for this?

**Review by p12tic — CHANGES_REQUESTED** [review]
Overall good, I just wrote a couple of nitpicks and it would be great to have a couple of missing trivial tests in the PR.

File: master/buildbot/test/unit/www/test_graphql.py (hunk 1 above)
Comment (tardyp) [reply master/buildbot/test/unit/www/test_graphql.py:70]: yes. for the moment, there is no output, just error management.

Comment (tardyp) [reply master/buildbot/test/unit/www/test_graphql.py:70]: the next PR will be the plumbing with data api. I guess it is easier for review to just land that part.

**Review by p12tic — APPROVED** [review]

**Pull request merged**